lightning/src/routing/network_graph.rs

```diff
@@ -202,6 +203,36 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
 	}
 }
 
+#[derive(PartialEq, Debug)]
+/// Tracks the score of a payment by holding onto certain metadata about
+/// the channel. Right now it stores a count of PaymentSent and PaymentFailed
+/// occurances for a given channel, but there's a lot more that could be
+/// added down the road (eg uptime, fee rates, variation, floppiness...)
+pub struct ChannelScore {
```

Comment:
I really think this needs to be abstract - as you point out, there are a lot of potential things that we could use here, and, more importantly, not all of them are going to be used in every case. eg, we might have a ProbingChannelScore-er which constantly probes the network, or we might only pull channel scores from failed/successful payments, as you do here, or a user might download channel scores from a server which provides some set of channel scoring information. Forcing users into a given set of score info isn't sufficient to capture what users may want.

Instead, we may consider generic-izing this to something which returns a "fee penalty" - to keep the route-finding algorithm tractible, the best thing we can do is have the score information be fed in as "fee we're willing to pay to avoid using this channel". Then, we can pass the `score_payment_failed_for_route`, etc calls into the trait, and have the router pull fee penalty information out of it.

Reply:
I just want to make sure I understand what you mean by generic-izing. Would the following be a valid approach to generalizing this to whatever a user may need?

Define a trait called `ChannelScorer` that requires a function to be defined that returns a `fee_penalty: u64`. Then, instead of the current `channel_score: ChannelScore` attribute in `DirectionalChannelInfo`, there would be an attribute called `channel_score: Option<T: impl ChannelScorer>`. Basically, any struct that can compute its own `fee_penalty` would be eligible to serve as that attribute. And if the user did not want to use that, it could just be a None. 

Concretely, a user could use the current `ChannelScore` for this. All the user would have to do is implement the required fee_penalty method and that could be used for route-finding. In the event that the user did not provide a struct for ChannelScorer, we could just go with the default route finding algorithm.